src/ui/colorLegendWidget.js

```diff
@@ -0,0 +1,421 @@
+      axisScale.exponent(category.exponent);
+    }
+    var id = uniqueID();
+    var precision = Math.max.apply(null, category.domain
+      .map(function (number) { return getPrecision(number); }));
+
+    var gradient = svg
+      .append('defs')
+      .append('linearGradient')
+      .attr('id', 'gradient' + id);
+    gradient.append('stop')
+      .attr('offset', '0%')
+      .attr('stop-color', category.colors[0]);
+    gradient.append('stop')
+      .attr('offset', '100%')
+      .attr('stop-color', category.colors[1]);
+    svg.append('rect')
+      .attr('fill', 'url(#gradient' + id + ')')
+      .attr('width', width)
+      .attr('height', '20px')
+      .on('mousemove', function () {
+        var value = axisScale.invert(d3.mouse(this)[0]);
+        var text = m_this._popupFormatter(value, precision);
+        m_this._showPopup(text);
+      })
+      .on('mouseout', m_this._hidePopup);
+
+    axis = d3.svg.axis()
+      .scale(axisScale)
+      .ticks(m_ticks, '.2s');
+
+    this._renderAxis(svg, axis);
+  };
+
+  /**
+   * Actually render the axis with d3.
+   * @param {Element} svg svg element that the axis will be drawn
+   * @param {object} axis d3 axis object
+   */
+  this._renderAxis = function (svg, axis) {
+    svg.append('g')
+      .attr('class', 'axis x')
+      .attr('transform', 'translate(0, 20)')
+      .call(function (g) {
+        g.call(axis);
+      });
+  };
+
+  /**
+   * Formatter of number that tries to maximize the precision
+   * while making the output shorter.
+   * @param {number} number to be formatted
+   * @param {precision} precision number of decimal precision will be tried to be kept
+   * @returns {string} formatted string output
+   */
+  this._popupFormatter = function (number, precision) {
+    number = parseFloat(number.toFixed(8));
+    precision = Math.min(precision, getPrecision(number));
+    precision = Math.min(precision, Math.max(3, 7 - Math.trunc(number).toString().length));
+    return d3.format('.' + precision + 'f')(number);
+  };
+
+  /**
+   * Show the popup based on current mouse event.
+   * @param {string} text content to be shown in the popup
+   */
+  this._showPopup = function (text) {
+    // The cursor location relative to the container
+    var offset = d3.mouse(m_this.canvas());
+    m_this.popup
+      .text(text);
+    var containerWidth = m_this.canvas().clientWidth;
+    var popupWidth = m_this.popup[0][0].clientWidth;
+    m_this.popup
+      .style({
+        // If the popup will be longer or almost longer than the container
+        'left': offset[0] - (offset[0] +
+          popupWidth - containerWidth > -10 ? popupWidth : 0) + 'px',
+        'top': (offset[1] - 22) + 'px'
+      })
+      .transition()
+      .duration(200)
+      .style('opacity', 1);
+  };
+
+  this._hidePopup = function () {
+    m_this.popup.transition()
+      .duration(200)
+      .style('opacity', 0);
+  };
+
+  return this;
+};
+
+function getPrecision(a) {
```

Comment:
This function needs to be commented (while you at it, _hidePopup, above, needs it too).